### PR TITLE
aws-lambda tutorial

### PR DESCRIPTION
**Pull request by lestephane**
This is the AWS Lambda tutorial for quantummaid, submitted for review.

I've re-added the .gitignore rules for all intellij assets, because I could not find anything in the checked in intellij files that needed checking in.

### CODE REVIEW

File: aws-lambda/01_TheCaseForLambda.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# The case for QuantumMaid + AWS Lambda
+
+## Why AWS Lambda?
+
+AWS Lambda has been a big deal since it was introduced back in 2014. With Lamba, there is no need to provision stateful
+servers such as docker containers or virtual machines instances. These incur hourly
```
Comment (quantummaiddeveloper) [inline aws-lambda/01_TheCaseForLambda.md]: - Docker needs to be capitalized (https://www.docker.com/legal/trademark-guidelines)
- virtual machines instances -> virtual machine instances

File: aws-lambda/02_MinimumViableFunction.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                        └── Main.java
+```
+
+## [`pom.xml`](step1/pom.xml)
+
+The most important `pom.xml` part is
+
+<!---[CodeSnippet](step1PomXml)-->
+```xml
+    <maven.compiler.parameters>true</maven.compiler.parameters><!-- ➊ -->
```
Comment (quantummaiddeveloper) [inline aws-lambda/02_MinimumViableFunction.md:26]: Starting the Maven snippet inside the `properties` tag seemed really strange. I just wanted to copy & paste it.

File: aws-lambda/02_MinimumViableFunction.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    <dependency>
+        <groupId>de.quantummaid.quantummaid.packagings</groupId>
+        <artifactId>quantummaid-essentials</artifactId>
+        <version>1.0.54</version>
+    </dependency>
+    <!--➋-->
+</dependencies>
+```
+
+➊ This is required to provide HttpMaid with method parameter name information through reflection. It's not enabled by default, so this must be explicitly enabled here.
```
Comment (quantummaiddeveloper) [inline aws-lambda/02_MinimumViableFunction.md]: We should not explicitly talk about HttpMaid at this point - it is sufficient to know that QuantumMaid needs method parameter name information.

File: aws-lambda/02_MinimumViableFunction.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+➋ This contains QuantumMaid reasonable defaults for dependencies, including minimal footprint alternatives to Jackson (minimal-json) and Logback (slf4j-simple). Because size matters.
```
Comment (quantummaiddeveloper) [inline aws-lambda/02_MinimumViableFunction.md:42]: I don't think we should tell the reader what alternative libraries could have been used (Jackson and Logback)

File: aws-lambda/02_MinimumViableFunction.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+## [`Main.java`](step1/src/main/java/de/quantummaid/tutorials/Main.java)
+
+Our starting point is the proverbial hello world from the [*"Getting Started"* guide](https://quantummaid.de/docs/01_gettingstarted.html)
```
Comment (quantummaiddeveloper) [inline aws-lambda/02_MinimumViableFunction.md]: Maybe `hello world` in italics? Or `hello-world`? Having the two words dangle around at the same context level as the rest of the text looks strange.

File: aws-lambda/03_AddingLambdaSupport.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
+# Step 2: Adding AWS Lambda support
+*(Full source code: [`step2`](step2) directory)*
+
+## Extracting HttpMaid's initialization code
+
+The HttpMaid initialization code will be exactly the same whether we run the code as a local HTTP endpoint or as an AWS Lambda function.
+
+In order for HttpMaid's initialization code to be shared between the local mode and the Lambda mode, we first need to extract it to a new method in the `Main` class:
```
Comment (quantummaiddeveloper) [inline aws-lambda/03_AddingLambdaSupport.md]: There is no HttpMaid code yet (should be "QuantumMaid")

File: aws-lambda/03_AddingLambdaSupport.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+private static QuantumMaid quantumMaidConfig() {
+  final QuantumMaid quantumMaid = QuantumMaid.quantumMaid()
+      .get("/helloworld", (request, response) -> response.setBody("Hello World!"));
+  return quantumMaid;
+}
+```
+
+## Adding the Lambda endpoint adapter
+
+Lambda integration is provided through an additional HttpMaid dependency:
```
Comment (quantummaiddeveloper) [inline aws-lambda/03_AddingLambdaSupport.md:20]: Maybe `additional HttpMaid dependency` -> `additional dependency`? It is in fact an HttpMaid dependency, but that it comes from HttpMaid doesn't seem to be important.

File: aws-lambda/03_AddingLambdaSupport.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+<!---[CodeSnippet](step2HttpMaidDependency)-->
+```xml
+<dependency>
+    <groupId>de.quantummaid.httpmaid.integrations</groupId>
+    <artifactId>httpmaid-awslambda</artifactId>
+    <version>0.9.72</version>
+</dependency>
+```
+
+Once the `httpmaid-lambda` dependency is added, a new class is available to bridge the HttpMaid world and the AWS Lambda world: `AwsLambdaEndpoint`.
```
Comment (quantummaiddeveloper) [inline aws-lambda/03_AddingLambdaSupport.md]: `HttpMaid world` -> `QuantumMaid world`

File: aws-lambda/03_AddingLambdaSupport.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+We should initialize an instance of `AwsLambdaEndpoint` in a static field of the `Main` class, so that:
+
+- The time taken to initialize HttpMaid does not count towards the execution time of the Lambda function. In plain english, it will not be billed by AWS.
```
Comment (quantummaiddeveloper) [inline aws-lambda/03_AddingLambdaSupport.md]: HttpMaid -> QuantumMaid
english -> English

File: aws-lambda/03_AddingLambdaSupport.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+➋ Use `Type: Api` if you want to use REST API instead of HTTP API. We use HTTP API because it's [_faster, lower cost and simpler to use_](https://aws.amazon.com/blogs/compute/building-better-apis-http-apis-now-generally-available/).
+
+➌➍ This means that requests to ➌ any path depth (`/`, `/helloworld`, `/hello/...`), using ➍ any method (`GET`, `HEAD`, `PUT`, `POST`, etc.), will be handled by our HttpMaid function. These parameters are fixed and required for a so-called [Lambda proxy integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html).
+
+Are we there yet? Almost.
+
+## Adding a public zero-argument constructor
+
+The AWS Lambda Java runtime instantiates the request handler class (`Main`) by calling it's default constructor,
```
Comment (quantummaiddeveloper) [inline aws-lambda/03_AddingLambdaSupport.md]: it's -> its

File: aws-lambda/05_UpdatingOurFunction.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
+# Step 4: Updating our function
+*(Full source code: [`step4`](step4) directory)*
+
+## /hello/`<whoever-you-are>`
+
+A function that returns `"Hello World!"` is not very useful.
+We'll make the returned Hello message [vary based on the URL path](https://quantummaid.de/docs/2_httpmaid/04_handlingrequests.html#request-route-and-path-parameters).
```
Comment (quantummaiddeveloper) [inline aws-lambda/05_UpdatingOurFunction.md]: Same thing as with `hello world`. Maybe

`"Hello"` message

or 

`"Hello"`-message

?

File: aws-lambda/05_UpdatingOurFunction.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+$ sam build
+$ sam deploy
+...
+Successfully created/updated stack - hello-app in us-east-1
+```
+
+## Viewing logs using SAM CLI
+
+Before we invoke the URL, let's tail the logs, which will give us some interesting details on our function execution(s).
+In a separate terminal, run:
```
Comment (quantummaiddeveloper) [inline aws-lambda/05_UpdatingOurFunction.md]: I missed the **separate** part. Maybe it should be written bold.

File: aws-lambda/05_UpdatingOurFunction.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+```shell
+$ time echo $(curl -s $apiUrl/hello/{me,you,everyone})
+Hello me!Hello you!Hello everyone!
+
+real	0m2.990s
+user	0m0.077s
+sys	0m0.005s
+```
+
+The logs display something similar to the following simplified output:
```
Comment (quantummaiddeveloper) [inline aws-lambda/05_UpdatingOurFunction.md]: Maybe it should be stated that it could take a bit of time until the logs arrive. This happened with me and I was confused.

File: aws-lambda/05_UpdatingOurFunction.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+- Time spent in Lambda = Init Duration: 870.58 ms + Duration: 398.08 ms + Duration: 3.32 ms + Duration: 3.32 ms = 1275.3 ms
+- Time spent overall = 2.990 s = 2990 ms
+- Unaccounted time = 2990 ms - 1275.3 ms = 1714.7 ms, of which
+  - approx. 600 ms is probably due to network latency (100 ms each way x 3 requests)
+  - approx. 1100 ms is as yet unexplained
+
+Having this log output at hand allows us to draw a few conclusions:
+
+- The HttpMaid overhead is flat beyond the first invocation (<5 ms)
```
Comment (quantummaiddeveloper) [inline aws-lambda/05_UpdatingOurFunction.md]: HttpMaid -> QuantumMaid
This also the only item in this list that does not end with a period

File: aws-lambda/05_UpdatingOurFunction.md [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+- The network latency is quite high here (200 ms per request). Maybe we can switch to an edge-optimized API.
+- If performance is important, 1100 ms is still unaccounted for and needs investigation. It could be TLS negotiation, or something else.
+- We can save money by reducing `MemorySize` in `template.yml` from 512 MB to something much closer to the maximum reported memory usage (101 MB), say, 128 MB.
+
+You now have the necessary building blocks to make your HttpMaid code AWS Lambda capable,
```
Comment (quantummaiddeveloper) [inline aws-lambda/05_UpdatingOurFunction.md:87]: HttpMaid -> QuantumMaid

File: aws-lambda/06_CleaningUp.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+Uh, the bucket is versioned, which means that we need to jump through additional hoops to empty it.
+A quick google search yields [this solution](https://stackoverflow.com/questions/29809105/how-do-i-delete-a-versioned-bucket-in-aws-s3-using-the-cli#35306665):
+
+```shell
+$ python3 -c "
+import boto3
+s3 = boto3.resource('s3')
+bucket = s3.Bucket('${sourceBucket}')
+bucket.object_versions.delete()
+"
```
Comment (quantummaiddeveloper) [inline aws-lambda/06_CleaningUp.md:54]: This did not work for me (something about "boto3 not found")

File: aws-lambda/itests/cf-lambda.yml [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+      MemorySize: 512
+      Tracing: Active
+      Events:
+        HelloWorldRestApi:
+          Type: Api
+          Properties:
+            RestApiId: !Ref HelloWorldRestApi
+            Path: /{proxy+}
+            Method: ANY
+        # HTTP Api does not support X-Ray (yet)
```
Comment (quantummaiddeveloper) [inline aws-lambda/itests/cf-lambda.yml]: Commented-out code

File: aws-lambda/itests/src/main/resources/README.md [hunk 18]
```diff
@@ -0,0 +1,7 @@
+This file's only purpose it to cause `mvn package` to create the target/classes.
```
Comment (quantummaiddeveloper) [inline aws-lambda/itests/src/main/resources/README.md]: it -> is

File: aws-lambda/step3/src/main/java/de/quantummaid/tutorials/Main.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+import static de.quantummaid.httpmaid.awslambda.AwsLambdaEndpoint.awsLambdaEndpointFor;
+
+public final class Main {
+  private static final AwsLambdaEndpoint ADAPTER = awsLambdaEndpointFor(quantumMaidConfig().httpMaid());
+
+  public Main() {}
+
+  public Map<String, Object> handleRequest(final Map<String, Object> request) {
+    System.err.println(request.toString());
```
Comment (quantummaiddeveloper) [inline aws-lambda/step3/src/main/java/de/quantummaid/tutorials/Main.java]: Is this supposed to be here?

File: aws-lambda/02_MinimumViableFunction.md (hunk 2 above)
Comment (lestephane) [reply aws-lambda/02_MinimumViableFunction.md:26]: I don't know how to hide the 

```
        <checkstyle.suppressions.location>
            ${project.basedir}/src/test/checkstyle/checkstyle-suppressions.xml
        </checkstyle.suppressions.location>
```

When inserting the full pom.xml file snippet. And I can't get the checkstyle checks to pass without the suppression setting. Catch 22.

File: aws-lambda/02_MinimumViableFunction.md (hunk 3 above)
Comment (lestephane) [reply aws-lambda/02_MinimumViableFunction.md]: done

File: aws-lambda/02_MinimumViableFunction.md (hunk 5 above)
Comment (lestephane) [reply aws-lambda/02_MinimumViableFunction.md]: Used `*Hello World*`

File: aws-lambda/06_CleaningUp.md (hunk 16 above)
Comment (lestephane) [reply aws-lambda/06_CleaningUp.md:54]: `pip install sam-aws-cli` pulls in the boto3 dependency. Could it be that you missed the `import boto3` statement? When I add `import boto3`, it finds it for me.

File: aws-lambda/05_UpdatingOurFunction.md (hunk 11 above)
Comment (lestephane) [reply aws-lambda/05_UpdatingOurFunction.md]: used `*Hello*`
